locator/token_metadata.cc

```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-    all_left_metadata->_impl->sort_tokens();
-}
-
 future<> token_metadata_impl::update_pending_ranges(
         const token_metadata& unpimplified_this,
         const abstract_replication_strategy& strategy, const sstring& keyspace_name, dc_rack_fn& get_dc_rack) {
     tlogger.debug("calculate_pending_ranges: keyspace_name={}, bootstrap_tokens={}, leaving nodes={}, replacing_endpoints={}",
         keyspace_name, _bootstrap_tokens, _leaving_endpoints, _replacing_endpoints);
     if (_bootstrap_tokens.empty() && _leaving_endpoints.empty() && _replacing_endpoints.empty()) {
         tlogger.debug("No bootstrapping, leaving nodes, replacing nodes -> empty pending ranges for {}", keyspace_name);
-        set_pending_ranges(keyspace_name, std::unordered_multimap<range<token>, inet_address>(), can_yield::no);
+        _pending_ranges_interval_map.erase(keyspace_name);
         return make_ready_future<>();
     }
 
     return async([this, &unpimplified_this, &strategy, keyspace_name, &get_dc_rack] () mutable {
-        std::unordered_multimap<range<token>, inet_address> new_pending_ranges;
-        calculate_pending_ranges_for_replacing(unpimplified_this, strategy, new_pending_ranges);
-        // Copy of metadata reflecting the situation after all leave operations are finished.
-        auto all_left_metadata = make_token_metadata_ptr(clone_after_all_left().get0());
-        calculate_pending_ranges_for_leaving(unpimplified_this, strategy, new_pending_ranges, all_left_metadata);
-        // At this stage newPendingRanges has been updated according to leave operations. We can
-        // now continue the calculation by checking bootstrapping nodes.
-        calculate_pending_ranges_for_bootstrap(strategy, new_pending_ranges, all_left_metadata, get_dc_rack);
-        all_left_metadata->clear_gently().get();
-
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
+            auto result = clone_only_token_map(false).get0();
+
+            // construct new_normal_tokens based on _bootstrap_tokens and _replacing_endpoints
+            std::unordered_map<inet_address, std::unordered_set<token>> new_normal_tokens;
+            if (!_replacing_endpoints.empty()) {
+                for (const auto& [token, inet_address]: _token_to_endpoint_map) {
+                    const auto it = _replacing_endpoints.find(inet_address);
+                    if (it == _replacing_endpoints.end()) {
+                        continue;
+                    }
+                    new_normal_tokens[it->second].insert(token);
+                }
+                for (const auto& [replace_from, replace_to]: _replacing_endpoints) {
```

Comment:
what's the point of replacing a node with itself? should it be allowed in the first place?

Reply:
It's entirely possible to replace a node that is down with a new node having the same ip address.
Fortunately, we changed it so the new node will have a new host_id. In the past it even inherited the replaced node's host_id which was even harder.

Reply:
ok, git it, thanks

Reply:
> tokens for replaced node will be added back in the loop "apply new_normal_tokens"

Right.

> the behaviour is changed though, previously pending_endpoints included the replaced endpoint, after the change it doesn't, this is demonstrated by the test. I'm not sure, is this important?

In the case of replacing with the same IP, you mean?

I think we should preserve the old behavior, because otherwise replacing node will not get the writes it's supposed to. The coordinator does this:

```
storage_proxy::create_write_response_handler_helper(...) {
    inet_address_vector_replica_set natural_endpoints = erm->get_natural_endpoints_without_node_being_replaced(token);
    inet_address_vector_topology_change pending_endpoints = erm->get_pending_endpoints(token, s->ks_name());
```

get_natural_endpoints_without_node_being_replaced() will filter-out the node which is being replaced. So the replacing node must be in pending endpoints for it to receive the writes.